File: pragya/apps/worker/src/handlers/predict.ts
import type { Handler } from "./index";

export const predictHandler: Handler = async (job, ctx) => {
  const participantId = String(job.payload.participant_id);
  const horizons = (job.payload.horizons_years as number[]) ?? [1, 3, 5];
  const seed = (job.payload.seed as number | undefined) ?? undefined;

  const features = await ctx.web.getParticipantFeatures(participantId);
  const modelId = (job.payload.model_id as string | undefined) ?? undefined;

  const result = (await ctx.ml.predict({
    participant_id: participantId,
    model_id: modelId,
    horizons_years: horizons,
    features: features.features,
    seed,
  })) as {
    model_id: string;
    model_version: string;
    predictions: Array<{
      horizon_years: number;
      risk_point: number;
      risk_lo_80: number;
      risk_hi_80: number;
      risk_lo_95: number;
      risk_hi_95: number;
    }>;
  };

  await ctx.web.writePrediction({
    participant_id: participantId,
    model_id: result.model_id,
    predictions: result.predictions,
    input_snapshot: features.features,
  });

  return {
    status: "succeeded",
    result: {
      model_id: result.model_id,
      model_version: result.model_version,
      horizons: result.predictions.map((p) => p.horizon_years),
    },
  };
};

File: pragya/apps/ml/drishti_ml/pipelines/predict.py
from __future__ import annotations

import hashlib

import numpy as np

from ..config import Settings
from ..models.registry import load_active_model
from .types import PredictRequest, PredictResponse, RiskInterval


def _stub_risk(participant_id: str, horizon: int, seed: int | None) -> float:
    h = hashlib.sha256(f"{participant_id}:{horizon}:{seed}".encode()).hexdigest()
    r = int(h[:8], 16) / 0xFFFFFFFF
    # Bias risk to grow with horizon.
    return float(min(0.95, max(0.01, r * 0.3 + (horizon / 10.0))))


def run(req: PredictRequest, settings: Settings) -> PredictResponse:
    active = load_active_model(settings)
    if active is None:
        # Phase 4 stub path.
        seed = req.seed if req.seed is not None else settings.default_seed
        predictions = []
        for h in req.horizons_years:
            p = _stub_risk(req.participant_id, h, seed)
            half80 = 0.05
            half95 = 0.10
            predictions.append(
                RiskInterval(
                    horizon_years=h,
                    risk_point=p,
                    risk_lo_80=max(0.0, p - half80),
                    risk_hi_80=min(1.0, p + half80),
                    risk_lo_95=max(0.0, p - half95),
                    risk_hi_95=min(1.0, p + half95),
                )
            )
        return PredictResponse(
            participant_id=req.participant_id,
            model_id="stub-v0",
            model_version="0.0.0",
            predictions=predictions,
        )

    # Phase 7: real model. The active model carries an ensemble of per-horizon classifiers.
    intervals = active.predict_intervals(req.features, req.horizons_years, seed=req.seed)
    return PredictResponse(
        participant_id=req.participant_id,
        model_id=active.model_id,
        model_version=active.version,
        predictions=[RiskInterval(**iv) for iv in intervals],
    )
